### PR TITLE
Consolidate user/group share actions into single dropdown

### PR DESCRIPTION
**Pull request by lordelix**
Rework user/group share actions. Concerning https://github.com/owncloud/enterprise/issues/2001

grafik

### CODE REVIEW

**Comment by felixheidecke** [comment]
@PVince81 I added the "silent model update" in order to prevent an immediate redraw wich closed the dropdown while changing stuff.

I hope that doesn't collide somewhere else. Tests are fine.

**Comment by felixheidecke** [comment]
I consider allowing the redraw on removing or adding the expiration date, so the little ⏱ immediately appears/disappears.

**Comment by phil-davis** [comment]
Seems to work nicely with my 15 minutes manual testing.

An unusual sequence that is odd is:
- enable, enforce and set the default/enforced days to 30 for user shares
- create a user share - it gets expiry date 30 days from now by default, good
- change the default/enforced days to 15 for user shares
- refresh the files page, redisplay the share you created. The expiry date shows as only 15 days in the future. (that might be "a good thing" if it really got reset to that)
- change the default/enforced days to 30 or more for user shares
- refresh the files page, redisplay the share you created. The expiry date shows as 30 days in the future. (which is the value that it really has in the server)

So when displaying an existing share, if the share has an expiry date past the current enforced limit, then the displayed expiry date is just the current enforced limit. (which is not a true indication of what is in the server database)

**Comment by phil-davis** [comment]
The UI tests will need refactoring for the new/changed behaviour here. Without having looked, I guess that it will not just be the Gherkin scenarios, but the placement of UI date elements has changed so xpath... in page objects probably needs sorting out.

**Comment by phil-davis** [comment]
I added some labels, because I guess QA-team will be helping with adjusting the UI acceptance tests.

**Comment by felixheidecke** [comment]
> refresh the files page, redisplay the share you created. The expiry date shows as only 15 days in the future. (that might be "a good thing" if it really got reset to that)

hmmm, I guess that's because the jQuery Datepicker sets a maxDate based on the enforced default expiration. Hmmm...

**Comment by phil-davis** [comment]
> hmmm, I guess that's because the jQuery Datepicker sets a maxDate based on the enforced default expiration. Hmmm...

I don't think it's a big thing - it only happens if the admin has set the default days to be long, then later reduces it.

**Comment by dpakach** [comment]
@felixheidecke, I just pushed a fix for the CI, Haven't run all the tests but they should be good now.

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 1]
```diff
@@ -103,15 +104,38 @@ private function getExpirationFieldFor($user, $type = 'user') {
 		return $this->find("xpath", \sprintf($this->shareWithExpirationFieldXpath, $user));
 	}
 
+	public function toggleShareActionsDropDown() {
```
Comment (micbar) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: Code Style complaining about missing doc comment.

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 2]
```diff
@@ -103,15 +104,38 @@ private function getExpirationFieldFor($user, $type = 'user') {
...
+		$this->assertElementNotNull(
+			$showCrudsBtn,
+			__METHOD__
+			. " xpath $this->showCrudsXpath could not find show-cruds button"
+		);
+		$showCrudsBtn->click();
+	}
+
+	public function openShareActionsDropDown() {
+		$this->toggleShareActionsDropDown();
```
Comment (micbar) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: same as above

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 3]
```diff
@@ -103,15 +104,38 @@ private function getExpirationFieldFor($user, $type = 'user') {
...
+		$this->waitTillElementIsNotNull($this->shareWithListDetailsXpath);
+	}
+
+	public function closeShareActionsDropDown() {
```
Comment (micbar) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: same as above

**Review by micbar — COMMENTED** [review]
Code Style

File: tests/acceptance/features/webUISharingInternalGroups/shareWithGroupUsingExpirationDate.feature [hunk 4]
```diff
@@ -25,15 +25,15 @@ Feature: Sharing files and folders with internal groups with expiration date set
     And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "no"
     And user "user1" has logged in using the webUI
     When the user shares file "lorem.txt" with group "grp1" using the webUI without closing the share dialog
-    Then the expiration date input field should not be visible for the group "grp1" in the share dialog
+    Then the expiration date input field should be visible for the group "grp1" in the share dialog
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalGroups/shareWithGroupUsingExpirationDate.feature]: We can actually do the sharing here, and taking the default "empty" expiration date. And then test that the share exists with no expiration date.

File: tests/acceptance/features/webUISharingInternalGroups/shareWithGroupUsingExpirationDate.feature [hunk 5]
```diff
@@ -25,15 +25,15 @@ Feature: Sharing files and folders with internal groups with expiration date set
...
 
   Scenario: expiration date is disabled for sharing with groups but enabled for sharing with users, user shares with a group
     Given user "user1" has uploaded file "filesForUpload/lorem.txt" to "/lorem.txt"
     And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "no"
     And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "yes"
     And user "user1" has logged in using the webUI
     When the user shares file "lorem.txt" with group "grp1" using the webUI without closing the share dialog
-    Then the expiration date input field should not be visible for the group "grp1" in the share dialog
+    Then the expiration date input field should be visible for the group "grp1" in the share dialog
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalGroups/shareWithGroupUsingExpirationDate.feature]: We can actually do the sharing here, and taking the default "empty" expiration date. And then test that the share exists with no expiration date.

File: tests/acceptance/features/webUISharingInternalUsers/shareWithUserUsingExpirationDate.feature [hunk 6]
```diff
@@ -16,15 +16,15 @@ Feature: Sharing files and folders with internal users with expiration date set/
     And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "no"
     And user "user1" has logged in using the webUI
     When the user shares file "lorem.txt" with user "User Two" using the webUI without closing the share dialog
-    Then the expiration date input field should not be visible for the user "User Two" in the share dialog
+    Then the expiration date input field should be visible for the user "User Two" in the share dialog
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalUsers/shareWithUserUsingExpirationDate.feature]: We can actually do the sharing here, and taking the default "empty" expiration date. And then test that the share exists with no expiration date.

File: tests/acceptance/features/webUISharingInternalUsers/shareWithUserUsingExpirationDate.feature [hunk 7]
```diff
@@ -16,15 +16,15 @@ Feature: Sharing files and folders with internal users with expiration date set/
...
 
   Scenario: expiration date is disabled for sharing with users but enabled for sharing with groups, user shares with an other user
     Given user "user1" has uploaded file "filesForUpload/lorem.txt" to "/lorem.txt"
     And parameter "shareapi_default_expire_date_user_share" of app "core" has been set to "no"
     And parameter "shareapi_default_expire_date_group_share" of app "core" has been set to "yes"
     And user "user1" has logged in using the webUI
     When the user shares file "lorem.txt" with user "User Two" using the webUI without closing the share dialog
-    Then the expiration date input field should not be visible for the user "User Two" in the share dialog
+    Then the expiration date input field should be visible for the user "User Two" in the share dialog
```
Comment (phil-davis) [inline tests/acceptance/features/webUISharingInternalUsers/shareWithUserUsingExpirationDate.feature]: We can actually do the sharing here, and taking the default "empty" expiration date. And then test that the share exists with no expiration date.

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 8]
```diff
@@ -139,13 +163,15 @@ public function getExpirationDateFor($user, $type = 'user') {
...
 	public function setExpirationDateFor(Session $session, $user, $type = 'user', $value = '') {
+		$this->openShareActionsDropDown();
 		$field = $this->getExpirationFieldFor($user, $type);
 		$this->assertElementNotNull(
 			$field,
 			"Could not find expiration field for " . $type . " '$user'"
 		);
 		$field->setValue($value . "\n");
 		$this->waitForAjaxCallsToStartAndFinish($session);
+		$this->closeShareActionsDropDown();
```
Comment (skshetry) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: I don't think, the page should be responsible for opening and closing the dropdown. This should probably be done from the context.

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 9]
```diff
@@ -103,15 +104,38 @@ private function getExpirationFieldFor($user, $type = 'user') {
...
 	 *
 	 * @return bool
 	 */
 	public function isExpirationFieldVisible($user, $type = 'user') {
+		$this->openShareActionsDropDown();
 		$field = $this->getExpirationFieldFor($user, $type);
-		return $field and $field->isVisible();
+		$visible = ($field and $field->isVisible());
+		$this->closeShareActionsDropDown();
+		return $visible;
```
Comment (skshetry) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: Same here

**Review by skshetry — COMMENTED** [review]
Just one thing, rest looked good to me. :)

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 10]
```diff
@@ -52,19 +52,20 @@ class SharingDialog extends OwncloudPage {
...
 	private $publicLinksTabContentXpath = "//div[@id='shareDialogLinkList']";
 	private $noSharingMessageXpath = "//div[@class='noSharingPlaceholder']";
 	private $publicLinkRemoveBtnXpath = "//div[contains(@class, 'removeLink')]";
 	private $publicLinkTitleXpath = "//span[@class='link-entry--title']";
 	private $notifyByEmailBtnXpath = "//input[@name='mailNotification']";
-	private $shareWithExpirationFieldXpath = "//*[@id='shareWithList']//span[@class='has-tooltip username' and .='%s']/..//input[contains(@class, 'expiration')]";
+	private $shareWithExpirationFieldXpath = "//*[@id='shareWithList']//span[@class='has-tooltip username' and .='%s']/../..//input[contains(@class, 'expiration')]";
 	private $shareWithClearExpirationFieldXpath = "/following-sibling::button[@class='removeExpiration']"; // in relation to $shareWithExpirationFieldXpath
 	private $shareWithListXpath = "//ul[@id='shareWithList']/li";
+	private $shareWithListDetailsXpath = "//div[@class='shareWithList__details']";
```
Comment (phil-davis) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: Does this class really have 2 `_` in its name?

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/core/22008/81/14
```
    Then the expiration date input field should be "+30 days" for the group "grp1" in the share dialog                                                                                  # WebUISharingContext::expirationDateShouldBe()
      Failed asserting that '15-01-2020' matches expected 1579083368.
```
in a few places.
Some problem with the `expirationDateShouldBe()` comparison code
@dpakach

File: tests/acceptance/features/bootstrap/WebUISharingContext.php [hunk 11]
```diff
@@ -255,10 +259,14 @@ public function expirationDateChangedTo($type, $receiver, $days) {
 	 * @param string $receiver
 	 *
 	 * @return void
+	 * @throws Exception
 	 */
 	public function expirationDateShouldBe($days, $type, $receiver) {
-		$expectedDate = \date('d-m-Y', \strtotime($days));
-		Assert::assertEquals($this->sharingDialog->getExpirationDateFor($receiver, $type), $expectedDate);
+		if (\strtotime($days) !== false) {
+			Assert::assertEquals(\strtotime($days), $this->sharingDialog->getExpirationDateFor($receiver, $type));
```
Comment (skshetry) [inline tests/acceptance/features/bootstrap/WebUISharingContext.php]:
```suggestion
			Assert::assertEquals(\date(\strtotime($days)), $this->sharingDialog->getExpirationDateFor($receiver, $type));
```

Probably, also format might need to be passed as `Y-m-d`.

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php (hunk 10 above)
Comment (felixheidecke) [reply tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: I'm trying to utilize the BEM method for CSS styles for better understandability:
https://css-tricks.com/bem-101/

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 12]
```diff
@@ -479,10 +479,10 @@ public function setSharingPermissions(
 				. " xpath $this->showCrudsXpath could not find show-cruds button for user "
 				. $shareReceiverName
 		);
+		$showCrudsBtn->click();
 		foreach ($permissions as $permission => $value) {
 			//the additional permission disappear again after they are changed
 			//so we need to open them again and again
```
Comment (phil-davis) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]: I guess this comment is no longer true?

**Comment by phil-davis** [comment]
This is failing: https://drone.owncloud.com/owncloud/core/22028/93/14

File: tests/acceptance/features/lib/FilesPageElement/SharingDialog.php [hunk 13]
```diff
@@ -103,6 +104,41 @@ private function getExpirationFieldFor($user, $type = 'user') {
...
+	 *
+	 * @return void
+	 */
+	public function openShareActionsDropDown() {
+		$this->toggleShareActionsDropDown();
+		$this->waitTillElementIsNotNull($this->shareWithListDetailsXpath);
+	}
+
+	/**
+	 * open the dropdown for share actions in the sidebar
```
Comment (phil-davis) [inline tests/acceptance/features/lib/FilesPageElement/SharingDialog.php]:
```suggestion
	 * close the dropdown for share actions in the sidebar
```
I will fix this after merging, when I am rebasing the PR to which this gets merged.

**Review by phil-davis — APPROVED** [review]

**Comment by phil-davis** [comment]
I will merge this to `feature/user-group-share-expiration` when CI goes green. Then I will rebase/squash/sort-out the commits in `feature/user-group-share-expiration`. Then we can do a "final" review there.

**Comment by dpakach** [comment]
@phil-davis one test is failing in drone but passing locally, I will try and find what's causing that.

**Comment by phil-davis** [comment]
> @phil-davis one test is failing in drone but passing locally, I will try and find what's causing that.

OK - if you have to push another commit, then you may as well also fix the test at SharingDialog.php line 133

**Review by phil-davis — APPROVED** [review]

**Comment by dpakach** [comment]
@felixheidecke `tests/acceptance/features/webUIRestrictSharing/restrictReSharing.feature:24` was failing in firefox. Currently we use older version of firefox (58.0). The dropdown doesn't works on that version but works on the recent one ( I tested firefox 71 and it worked). Does the older version needs to be supported?

**Comment by felixheidecke** [comment]
> @felixheidecke `tests/acceptance/features/webUIRestrictSharing/restrictReSharing.feature:24` was failing in firefox. Currently, we use older version of firefox (58.0). The dropdown doesn't works on that version but works on the recent one ( I tested firefox 71 and it worked). Does the older version needs to be supported?

@dpakach Looks like @phil-davis decided that it doesn't need to be supported ;-)

**Comment by phil-davis** [comment]
Yes, we talked about it. We manually tested in the latest Firefox and it works. We do not support every version of Firefox - so it would be "a bit random" to insist on fixing something on Firefox58 and other things might or might not work on Firefox42,57,59,60,61,...

**Pull request merged**